text/3634-scoped-impl-trait-for-type.md

```diff
@@ -0,0 +1,3436 @@
+
+\*\* but see [which-structs-should-be-implementation-invariant].
+
+See also [why-specific-implementation-invariant-generics].
+
+[call expressions]: https://doc.rust-lang.org/reference/expressions/call-expr.html#call-expressions
+
+## Call expressions' *function operand* captures its *implementation environment*
+[call-expressions-function-operand-captures-its-implementation-environment]: #call-expressions-function-operand-captures-its-implementation-environment
+
+Call expressions capture the *implementation environment* in their *function operand*, acting as host for [implementation-invariant-generics]. This enables call expressions such as
+
+```rust
+Option::<Type as Debug in module>::fmt(…)
+```
+
+where `fmt` receives the specified scoped implementation by observing it through the `T: Debug` bound on its implementing `impl` block.
+
+If no observing bound exists, code of this form should produce a warning spanning the `Trait in module` tokens. (see [unused-scoped-implementation])
+
+## Type aliases are opaque to scoped implementations
+[type-aliases-are-opaque-to-scoped-implementations]: #type-aliases-are-opaque-to-scoped-implementations
+
+As scoped `impl Trait for Type` is a fully lexically-scoped feature, the *implementation environment* present in a scope does not affect types hidden behind a type alias, except for the top-level type directly:
+
+```rust
+trait Trait {
+    fn method(&self) -> &str;
+}
+
+impl Trait for Type {
+    fn method(&self) -> &str {
+        "global"
+    }
+}
+
+mod m1 {
+    use super::Type;
+
+    pub type Alias = [Type; 1];
+}
+
+mod m2 {
+    use super::{Type, Trait};
+
+    pub use impl Trait for Type {
+        fn method(&self) -> &str {
+            "scoped"
+        }
+    }
+
+    pub use impl<T: Trait> Trait for [T; 1] {
+        fn method(&self) -> &str {
+            self[0].method()
+        }
+    }
+}
+
+fn main() {
+    use m1::Alias;
+    use m2::{
+        impl Trait for Type,
+        impl Trait for [Type; 1],
+    };
+
+    assert_eq!([Type].method(), "scoped");
+    assert_eq!(Alias::default().method(), "global");
+}
+```
+
+Scoped implementations may still be observed through bounded generic type parameters on the type alias itself. (see [binding-choice-by-implementations-bounds])
+
+## `TypeId` of generic type parameters' opaque types
+[typeid-of-generic-type-parameters-opaque-types]: #typeid-of-generic-type-parameters-opaque-types
+
+In addition to the type identity of the specified type, the `TypeId` of opaque generic type parameter types varies according to the captured *implementation environment*, but *only according to implementations that are relevant to their bounds (including implicit bounds)*, so that the following program runs without panic:
+
```

Comment:
I could word this better. The type of generic type parameters is always opaque, as it abstracts the various concrete types inserted in its place during monomorphisation (but also, due to implementation environment "baking", these concrete types can be effectively inexpressible in Rust).

```suggestion
## `TypeId` of generic type parameters' opaque types
[typeid-of-generic-type-parameters-opaque-types]: #typeid-of-generic-type-parameters-opaque-types

In addition to the type identity of the specified type, the `TypeId` of generic type parameter types varies according to the *implementation environment* that has been captured into them, but *only according to implementations that are relevant to their bounds (including implicit bounds)*, so that the following program runs without panic:
```

Reply:
Resolved in f3589adf495ceb828627817e234f4dffdaedc5b8, hopefully, but feel free to unresolve if it's still unclear.